messageBundlesCompiler.js

```diff
@@ -0,0 +1,198 @@
+/**
+ * The message bundles compiler.
+ *
+ * Contains various functions for compiling a mega message bundle which contains
+ * all messages for all available locales.
+ * Copyright 2017 Raising the Floor - International
+ *
+ * Licensed under the New BSD license. You may not use this file except in
+ * compliance with this License.
+ * The research leading to these results has received funding from the European Union's
+ * Seventh Framework Programme (FP7/2007-2013) under grant agreement no. 289016.
+ * You may obtain a copy of the License at
+ * https://github.com/GPII/universal/blob/master/LICENSE.txt
+ */
+
+"use strict";
+
+var fs = require("fs");
+var path = require("path");
+
+var DEFAULT_FILE_TYPE = "json",
+    DEFAULT_PARSER = JSON;
+
+/**
+ * Generates a list of the least to the most specific locales that can be used.
+ * @param locale {String} A locale, e.g. "en_us".
+ * @param defaultLocale {String} The locale to be used in case messages from all
+ * previous locales are missing.
+ * @returns {String[]} A list of locales. For example, for the "fr_FR" locale and
+ * in case the default locale is "en", the function would return ["en", "fr", "fr_FR"].
+ */
+function getLocaleVersions(locale, defaultLocale) {
+    var segments = locale.split("_");
+
+    var versions = segments.map(function (segment, index) {
+        return segments.slice(0, index + 1).join("_");
+    });
+
+    if (defaultLocale) {
+        versions.unshift(defaultLocale);
+    }
+
+    return versions;
+}
+
+/**
+ * In case there are only partial messages for a given locale, this function adds
+ * the missing messages using the messages for all versions of the locale in the
+ * order in which they are returned by the `getLocaleVersions` function.
+ * @param messageBundles {Object} A raw map of collected bundles. It is of type
+ * { locale1: { key1: message1, ... }, locale2: { key2: message2 }, ... }
+ * @param defaultLocale {String} The default locale to be used.
+ * @returns {Object} The enhanced message bundles object.
```

Comment:
Use a more specific type here such as {MessageBundles} and define its structure in a mega-comment somewhere